crates/typst-cli/src/update.rs

```diff
@@ -0,0 +1,359 @@
+
+/// Figure out if there are sufficient permissions to carry out an update --
+/// hard failing if the cli is installed through a package manager.
+#[derive(Clone, Copy, Debug)]
+enum SelfUpdatePermission {
+    HardFail,
+    Permit,
+}
+
+/// A GitHub release.
+#[derive(Debug, Deserialize)]
+struct Release {
+    name: String,
+    tag_name: String,
+    assets: Vec<Asset>,
+}
+
+/// Assets that were uploaded to a GitHub release.
+///
+/// Primarly used to download pre-compiled typst CLI binaries.
+#[derive(Debug, Deserialize)]
+struct Asset {
+    pub name: String,
+    pub browser_download_url: String,
+}
+
+/// GitHub asset archive with the typst CLI executable as unpacked data.
+#[derive(Debug)]
+struct Archive {
+    pub extension: Extension,
+    pub buffer: BufReader<Cursor<Vec<u8>>>,
+}
+
+/// The extension for the downloaded archive.
+///
+/// Possible variations are `zip` and `xz`, other variations are mapped to
+/// `Unsupported` and will throw an error.
+#[derive(Debug)]
+enum Extension {
+    Zip,
+    Xz,
+    Unsupported,
+}
+
+impl From<&str> for Extension {
+    fn from(value: &str) -> Self {
+        match value {
+            "zip" => Extension::Zip,
+            "xz" => Extension::Xz,
+            _ => Extension::Unsupported,
+        }
+    }
+}
+
+/// Self update the typst CLI binary.
+///
+/// Fetches a target release or the latest release (if no version was specified)
+/// from GitHub, unpacks it and self replaces the current binary with the
+/// pre-compiled asset from the downloaded release.
+pub fn update(command: UpdateCommand) -> StrResult<()> {
+    let update_permitted = if NEVER_SELF_UPDATE {
+        SelfUpdatePermission::HardFail
+    } else {
+        self_update_permitted()?
+    };
+
+    match update_permitted {
+        SelfUpdatePermission::HardFail => {
+            println!("self-update is disabled for this build of the typst cli");
+            println!(
+                "you should probably use your system package manager to update typst"
+            );
+            // not really an ok scenario but not really an error either?
+            return Ok(());
+        }
+        SelfUpdatePermission::Permit => {}
+    }
+
+    // first we check if a downgrade is happening
+    if let Some(ref version) = command.version {
+        let current_tag = env!("CARGO_PKG_VERSION").parse().unwrap();
+        if !command.force && version < &current_tag {
+            println!("Certain downgraded typst versions will not have the update command available");
+            println!("Forcing a downgrade might break your install");
+            println!("You can downgrade by running `typst update <VERSION> --force`");
+
+            return Ok(());
+        }
+    }
+
+    let executable = env::current_exe()
+        .map_err(|err| eco_format!("failed to grab current exe path: {}", err))?;
+    let backup = executable
+        .parent()
+        .unwrap_or(Path::new("./"))
+        .join("typst_backup.part");
```

Comment:
Agreed /usr/bin is not a good place to leave this, but a user specific XDG cache dir might not be right either. If backing up a system level install process the backup should also be in a system level location. A sys admin coming along later should see what the state is even if a user with sudo got themselves in a pickle without looking in their $HOME.

Reply:
> Agreed /usr/bin is not a good place to leave this, but a user specific XDG cache dir might not be right either. If backing up a system level install process the backup should also be in a system level location. A sys admin coming along later should see what the state is even if a user with sudo got themselves in a pickle without looking in their $HOME.

Note that `typst update` wouldn't be able to affect a system install without `sudo`, so it wouldn't be a sys admin operation. If `sudo typst update` or something like that is run, then it would likely go into `/root`'s cache dir as a consequence (which would be easily accessible by a sys admin), although we could also add some `/opt` override for that specific case.

Reply:
Sys admin stuff like binaries and rollback data never goes in /root any more than anywhere in /home, that's just a home directory for a particular user. And no, using sudo does not always set the home directory to root's either. And no, /opt is not going to be the right place on many Linux systems either.

Reply:
Well, what are you thinking of then? `/var`?

(I guess we could also allow for an environment variable to override whatever we choose.)

Reply:
Yes, having a quick look at [FHS](https://refspecs.linuxfoundation.org/FHS_3.0/fhs/ch05s08.html#purpose38) the best I've thought of so far would be `/var/lib/typst/`.